rust/lancedb/src/arrow.rs

```diff
@@ -119,3 +128,214 @@ impl<T: arrow_array::RecordBatchReader + Send + 'static> IntoArrow for T {
+            index: 0,
+        }
+    }
+}
+
+#[cfg(feature = "polars")]
+impl Iterator for PolarsDataFrameRecordBatchReader {
+    type Item = std::result::Result<arrow_array::RecordBatch, arrow_schema::ArrowError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index < self.chunks.len() {
+            let columns: Vec<arrow_array::ArrayRef> = self.chunks[self.index]
+                .arrays()
+                .iter()
+                .map(|polars_array| arrow_array::ArrayRef::from(&**polars_array))
+                .collect();
+            self.index += 1;
+            Some(arrow_array::RecordBatch::try_new(
+                self.arrow_schema.clone(),
+                columns,
+            ))
+        } else {
+            None
+        }
+    }
+}
+
+#[cfg(feature = "polars")]
+impl arrow_array::RecordBatchReader for PolarsDataFrameRecordBatchReader {
+    fn schema(&self) -> Arc<arrow_schema::Schema> {
+        self.arrow_schema.clone()
+    }
+
+    fn next_batch(
+        &mut self,
+    ) -> std::prelude::v1::Result<Option<arrow_array::RecordBatch>, arrow_schema::ArrowError> {
+        self.next().transpose()
+    }
+}
+
+/// A trait for converting the result of a LanceDB query into a Polars DataFrame with aligned
+/// chunks. The resulting Polars DataFrame will have aligned chunks, but the series's
+/// chunks are not guaranteed to be contiguous.
+#[cfg(feature = "polars")]
+pub trait IntoPolars {
+    fn into_polars(self) -> impl std::future::Future<Output = Result<DataFrame>> + Send;
+}
+
+#[cfg(feature = "polars")]
+impl IntoPolars for SendableRecordBatchStream {
+    async fn into_polars(self) -> Result<DataFrame> {
+        let arrow_schema = self.schema();
+        let polars_schema = convert_arrow_schema_to_polars_schema(&arrow_schema);
+        let mut acc_df: DataFrame = DataFrame::from(&polars_schema);
+        let record_batches = self.try_collect::<Vec<_>>().await?;
+        for record_batch in record_batches {
+            let new_df = convert_record_batch_to_polars_df(&record_batch, &polars_schema)?;
+            acc_df = acc_df.vstack(&new_df)?;
+        }
+        return Ok(acc_df);
+    }
+}
+
+#[cfg(feature = "polars")]
+fn convert_arrow_schema_to_polars_schema(arrow_schema: &arrow_schema::Schema) -> Schema {
+    Schema::from_iter(arrow_schema.fields().iter().map(|field| {
+        Field::new(
+            field.name(),
+            datatypes::DataType::from(&datatypes::ArrowDataType::from(field.data_type().clone())),
+        )
+    }))
+}
+
+#[cfg(feature = "polars")]
+fn convert_record_batch_to_polars_df(
+    record_batch: &arrow::record_batch::RecordBatch,
+    polars_schema: &Schema,
+) -> Result<DataFrame> {
+    let mut columns: Vec<Series> = Vec::with_capacity(record_batch.num_columns());
+
+    for (i, column) in record_batch.columns().iter().enumerate() {
+        let polars_array = Box::<dyn array::Array>::from(&**column);
```

Comment:
`&**` feels weird but maybe it's inevitable.

Reply:
Yeah, I wasn't able to find a way around it.